### ISSUE TITLE
Incorrect type inference for table field on union types

### ISSUE THREAD

**Issue by Rinkaa** [EmmyLuaLs/emmylua-analyzer-rust#796]
Example code:

```lua
---@type {woof:(fun():integer), meow:nil} | {meow:(fun():string)}
local t = {woof = function() return 1 end}
if t.meow then
  print("cat")
end
```

<img width="583" height="121" alt="Image" src="https://github.com/user-attachments/assets/a9fcbd46-4068-4524-b1fd-274377e74943" />

The code is for checking whether a field of a table exists, in order to decide its subtype in the union type and go through different processes accordingly. 
Type reference of `t.meow` is `fun()->string`.
It results in a false warning "Unnecessary `if` statement: condition is always true".

有时需要检查一个table中某成员是否存在来辨别是多个类型中的哪个类型，做对应处理。当前的类型推断是认为只要成员出现在任何子类型中都算作存在。想问问这个算符合预期设计还是bug？感觉有点奇怪……

In Lua-LS, the type of `t.meow` is `fun():string|nil`.

Possibly related to #717

**Comment by Rinkaa** [comment]
@CppCXY  可能还要reopen一下这个issue（我发现我没有reopen权限），因为在变更之后出现了关联的新的问题
```lua
---@type {woof:(fun():integer), meow:nil} | {meow:(fun():string), wait?:(fun(time:number))}
local t = {woof = function() return 1 end}
if t.meow then
  if t.wait then
    t.wait(1)
  end
end
```

<img width="697" height="217" alt="Image" src="https://github.com/user-attachments/assets/e2b44c1a-e298-438f-934f-2687bd9cf571" />

在先前原有的代码增加了一个可空成员`wait?`之后，`t.wait`的类型
- 在第4行`if t.wait then`处变成了`(fun(...) -> ...)??`
- 在第5行`t.wait(1)`处变成了`(fun(time: number))?`

但在lua层面来说，有了if判断真值，在if内t.wait就不可能为空了。感觉类型推断还是出了点情况
